feat: move Identify I/O from NetworkBehaviour to ConnectionHandler

jxs opened the pull request:
## Description
Addresses #2885 
<!-- Please write a summary of your changes and why you made them.-->
<!-- This section will appear as the commit message after merging. Please craft it accordingly. -->

## Notes
Following the discussion on https://github.com/libp2p/rust-libp2p/issues/2885 this was the simplest way I found to move the I/O to `Handler`. 
Per commit review is suggested. 
## Links to any relevant issues

<!-- Reference any related issues.-->

## Open Questions

<!-- Unresolved questions, if any. -->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

jxs committed:
Identify: rename pending IdentifyHandler references on doc.

jxs committed:
Identify: start moving I/O from NetworkBehaviour,

instead of responding pending replies from NetworkBehaviour, send them
back to ConnectionHandler.

ConnectionHandler for now just receives them, it's implementation of the
responding will come next.

jxs committed:
identify: Move I/O from Networkehaviour,

reply to the Identification requests from the ConnectionHandler.

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -63,18 +61,10 @@ pub struct Behaviour {
-        io: Pin<Box<dyn Future<Output = Result<(), UpgradeError>> + Send>>,
-    },
+struct Request {
+    peer: PeerId,
```
The handler could learn this via `IntoConnectionHandler`. It will never change across its lifespan.

jxs replied:
updated!

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -63,18 +61,10 @@ pub struct Behaviour {
-    },
+struct Request {
+    peer: PeerId,
+    io: ReplySubstream<NegotiatedSubstream>,
```
It would be nice to not pass this back and forth.

jxs replied:
Yeah agreed thanks, updated!

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -63,18 +61,10 @@ pub struct Behaviour {
+struct Request {
+    peer: PeerId,
+    io: ReplySubstream<NegotiatedSubstream>,
+    observed: Multiaddr,
```
Same here, a `Handler` can learn this via `IntoConnectionHandler` and it won't change[^1] across its lifetime.

[^1]: `AddressChange` is not emitted at the moment but once it will, it is directly given to the handler.

jxs replied:
but we only get it on `connection_established` which seems to be called after `new_handler`.

thomaseizinger replied:
You have to implement `IntoConnectionHandler` additionally on a struct that you return from `new_handler`, said struct can then capture the `ConnectedPoint` and initialise the `ConnectionHandler` with it.

jxs replied:
thanks Thomas, updated!

thomaseizinger reviewed (commented):
Thanks!

It would be nice if we changed a few more things:

- Initialise the `ConnectionHandler` with `PeerId` and observed address directly
- Don't send the substream back and forth
- Upon receiving a new substream, send an event to the behaviour asking for more information like protocols, upon answer, grab the next substream in line and send it back

I think this would be an overall easier design because it clearly shows, what information is needed from the behaviour that can't be accessed by the handler.

We can even pass stuff like the public key and the agent version directly to the handler upon initialisation.

jxs committed:
review: Don't bubble the Substream from the handler

to the behaviour, instead keep track of the info as it doesn't change.

jxs committed:
review: improve Request doc wording.

jxs commented on protocols/identify/src/handler.rs:
```diff
@@ -255,6 +302,46 @@ impl ConnectionHandler for Handler {
         }
 
+        // Check for pending replies to send.
+        if let Some(ref info) = self.info {
```
NB: I thought `let_chains` had landed already been stabilized, but[ it was reverted](https://github.com/rust-lang/rust/pull/100538).

thomaseizinger replied:
Really? Did you try with Rust 1.65?

jxs replied:
yeah, compiler points me to https://github.com/rust-lang/rust/issues/53667, see https://github.com/rust-lang/rust/issues/53667#issuecomment-1297933589

jxs reviewed (commented):
Thanks for the fast review and guidance Thomas!

>I think this would be an overall easier design because it clearly shows, what information is needed from the behaviour that can't be accessed by the handler.
>
>We can even pass stuff like the public key and the agent version directly to the handler upon initialisation.

the downside of that is that information comes from multiple different places instead of coming only from a single `Info` struct. I updated to code to no longer send back and forth the substream, if you still feel that we should pass first the `PeerId` on constructor and then all the other details when requested, to prepare for the changes discussed on https://github.com/libp2p/rust-libp2p/issues/2885 no worries I'll update the PR :)

jxs committed:
review: separate Info details provided by the behaviour,

provided on new_handler constructor Everything that doesn't change over
the lifetime.

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into identify-remove-io-behaviour

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -51,8 +48,8 @@ pub struct Behaviour {
-    /// Pending replies to send.
-    pending_replies: VecDeque<Reply>,
+    /// Information requests from the handlers to be fullfiled.
+    requests: VecDeque<PeerId>,
```
I think you'll need to track the connection ID here so you can respond to the correct handler!

thomaseizinger commented on protocols/identify/src/behaviour.rs:
```diff
@@ -346,55 +327,17 @@ impl NetworkBehaviour for Behaviour {
+            };
+            return Poll::Ready(NetworkBehaviourAction::NotifyHandler {
+                peer_id: peer,
+                handler: NotifyHandler::Any,
```
Here is the curlpit! This will be buggy with > 1 connection per peer.

jxs replied:
Thanks Thomas, addressed.

thomaseizinger commented on protocols/identify/src/handler.rs:
```diff
@@ -90,36 +130,84 @@ pub struct Handler {
+    observed_addr: Multiaddr,
+
+    /// Information provided by the `Behaviour` upon requesting.
+    behaviour_info: Option<BehaviourInfo>,
```
Instead of tracking this here, I think it would be cleaner to:

1. Store pending substreams in a queue
2. Request a new `BehaviourInfo` every time a new substream comes in
3. Construct a `Sending` every time you receive a `BehaviourInfo` from the behaviour by popping the oldest substream from the queue in (1)

That will work off substreams in a FIFO fashion and feed them the most up-to-date information.

thomaseizinger replied:
Alternatively, we could initialise this in the constructor as well. I think we can assume that `new_handler` in the behaviour only gets called after it got polled at least once!

jxs replied:
> Instead of tracking this here, I think it would be cleaner to:
> 
> 1. Store pending substreams in a queue
> 2. Request a new `BehaviourInfo` every time a new substream comes in
> 3. Construct a `Sending` every time you receive a `BehaviourInfo` from the behaviour by popping the oldest substream from the queue in (1)
> 
> That will work off substreams in a FIFO fashion and feed them the most up-to-date information.

Thanks, makes sense to me. But used two queues instead of one with different states for a less confusing management of them ptal Thomas.

thomaseizinger commented on protocols/identify/src/handler.rs:
```diff
@@ -90,36 +130,84 @@ pub struct Handler {
+
+/// Information provided by the `Behaviour` upon requesting.
+#[derive(Debug)]
+pub struct BehaviourInfo {
```
Not the best name but should do for now I guess :)

thomaseizinger reviewed (commented):
Thanks! A few more ideas!

jxs committed:
review: handler, request info on each new substream,

instead of caching the behaviour info.

thomaseizinger commented on protocols/identify/src/handler.rs:
```diff
@@ -207,14 +299,37 @@ impl ConnectionHandler for Handler {
+                    .reply_streams
+                    .pop_front()
+                    .expect("A BehaviourInfo reply should have a matching substream.");
+                let io = Box::pin(substream.send(info));
```
```suggestion
                let io = Box::pin(async move { substream.send(info).await?; Ok(self.remote_peer_id) });
```

If you construct the future this way, you don't need the `Sending` struct and can you `FuturesUnordered` instead :)

jxs replied:
ah! How nice, thanks Thomas! :)

thomaseizinger commented on protocols/identify/src/handler.rs:
```diff
@@ -90,36 +128,81 @@ pub struct Handler {
+    /// Identifying information of the local node that is pushed to a remote.
+    Push(Info),
+    /// Identifying information requested from this node.
+    Identify(BehaviourInfo),
```
How different are these two really? Isn't exactly the same, just with different semantics?

jxs replied:
thanks for recalling this again
```
large size difference between variants
   --> protocols/identify/src/handler.rs:174:1
    |
174 | / pub enum InEvent {
175 | |     /// Identifying information of the local node that is pushed to a remote.
176 | |     Push(Info),
    | |     ---------- the largest variant contains at least 304 bytes
177 | |     /// Identifying information requested from this node.
178 | |     Identify(BehaviourInfo),
    | |     ----------------------- the second-largest variant contains at least 48 bytes
179 | | }
    | |_^ the entire enum is at least 304 bytes
    |
    = note: `-D clippy::large-enum-variant` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
help: consider boxing the large fields to reduce the total size of the enum
    |
176 |     Push(Box<Info>),
    |          ~~~~~~~~~
```

thomaseizinger replied:
I wasn't talking about the size difference actually.

Semantically, even for an identify push, all the handler needs to know are the updated protocols and listen addresses, which is represented as the `BehaviourInfo` struct. Can we model it as the same type?

jxs replied:
ah right! So I followed along and merge `Behaviour`'s `pending_push` into `requests` for better balance when fullfiling them. Previously `Push` requests would always have priority over `Identify` ptal Thomas.

thomaseizinger commented on protocols/identify/src/handler.rs:
```diff
@@ -82,6 +114,12 @@ pub struct Handler {
     >,
 
+    /// Streams awaiting `BehaviourInfo` to then send identify requests.
+    reply_streams: VecDeque<ReplySubstream<NegotiatedSubstream>>,
```
Is the abstraction of a `ReplySubstream` useful here? Would it be easier to just buffer the streams themselves?

jxs replied:
Sorry Thomas, not following. What is the alternative? But the `send` method is implemented for `ReplySubstream`.

thomaseizinger replied:
The alternative is inlining `send` to wherever it is called which I'd guess is only one place?

thomaseizinger replied:
We don't really do this wrapping of a substream anywhere else in the codebase so it would be nice to align it here.

Constructing a `Framed` is what we typically do.

jxs replied:
ah, right! Thanks! I was able to remove `ReplySubstream` but `send` is still being used in more places. 
We can go further and deprecate `Info` and just use the protobuf `Identify`, but probably do that on a subsequent PR?

thomaseizinger replied:
> ah, right! Thanks! I was able to remove `ReplySubstream` but `send` is still being used in more places. 
> We can go further and deprecate `Info` and just use the protobuf `Identify`, but probably do that on a subsequent PR?

Nah, I think that is fine. We usually don't use the protobuf struct directly which is good.

thomaseizinger reviewed (approved):
Thanks! This looks good to me now, one more idea but happy to merge anytime!

jxs committed:
review: use FuturesUnordered for pending replies.

jxs committed:
review: rename Protocol and ProtocolPush to Identify and Push.

jxs committed:
review: merge BehaviorInfo and InEvent,

into a single struct with the corresponding Substream protocol.

jxs committed:
review: merge Behaviour pending_push into requests,

to better balance incoming requests and fullfil them by order.

jxs committed:
review: deprecate ReplySubstream.

thomaseizinger reviewed (approved):
Very nice, this actually went beyond what I thought could be simplified :)

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into identify-remove-io-behaviour

jxs committed:
review: fix doc typos.

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into identify-remove-io-behaviour

mxinden reviewed (commented):
Great simplifications along the main goal of the pull request. Thanks.

Needs a changelog entry. Otherwise good to go from my end.

jxs committed:
review: add CHANGELOG.md entry

jxs committed:
Merge branch 'master' of github.com:libp2p/rust-libp2p into identify-remove-io-behaviour

mxinden reviewed (approved):
:pray:
